### PR TITLE
Featured Image of a page

### PR DESCRIPTION
**Pull request by jckautzmann**
This PR adds following features to the Page, Image, Teaser and List components:
- page dialog: supports adding a featured image to a page
- image component: uses the featured image of the page when no image is defined
- teaser component: uses the featured image of the linked page (if available) when no image is defined
- list component: displays a list of teasers based on the featured image of the linked pages

It also relates to #770.

Inheritance works as follows:
- image: all the properties are retrieved from the featured image, except for the resource type and the ID, which are retrieved from the caller image.
- teaser: the teaser model creates a wrapped image resource based on the feature image of the target page and adds the page path as link URL to it. The teaser specific properties are  retrieved from the teaser.
- list: the list model creates a wrapped teaser resource based on the feature image of the target page and adds the page title, page description and page link to it. Note that setting the `Item Settings` of the dialog is currently not working.

### CODE REVIEW

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v2/PageListItemImpl.java [hunk 1]
```diff
@@ -15,21 +15,41 @@
...
+import com.adobe.cq.wcm.core.components.util.ComponentUtils;
+import com.day.cq.commons.ImageResource;
+import com.day.cq.commons.jcr.JcrConstants;
 import com.day.cq.wcm.api.Page;
 import com.day.cq.wcm.api.components.Component;
 import com.fasterxml.jackson.annotation.JsonIgnore;
 
 public class PageListItemImpl extends com.adobe.cq.wcm.core.components.internal.models.v1.PageListItemImpl {
 
+    public static final String TEASER_DELEGATE = "teaserDelegate";
```
Comment (vladbailescu) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v2/PageListItemImpl.java:40]: I would rather we move this constant to the interface of the component where the property is used (ie `List`)

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/TeaserResourceWrapper.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import org.apache.sling.api.resource.ResourceWrapper;
+import org.apache.sling.api.resource.ValueMap;
+import org.apache.sling.api.wrappers.ValueMapDecorator;
+import org.apache.sling.models.annotations.Exporter;
+import org.jetbrains.annotations.NotNull;
+
+import com.adobe.cq.export.json.ExporterConstants;
+
+@Exporter(name = ExporterConstants.SLING_MODEL_EXPORTER_NAME , extensions = ExporterConstants.SLING_MODEL_EXTENSION)
+public class TeaserResourceWrapper extends ResourceWrapper {
```
Comment (vladbailescu) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/TeaserResourceWrapper.java:33]: This isn't teaser specific and could be used with any resource to change the RT and properties.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/TeaserResourceWrapper.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    private ValueMap valueMap;
+    private String resourceType;
+
+    public TeaserResourceWrapper(@NotNull Resource resource, @NotNull String resourceType, @NotNull Map<String, String> inheritedProperties) {
```
Comment (vladbailescu) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/TeaserResourceWrapper.java:38]: I'd rename `inheritedProperties` to `overriddenProperties` and `resourceType` to `overriddenResourceType`

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/AbstractImageDelegatingModel.java [hunk 4]
```diff
@@ -42,11 +44,18 @@
...
 
     protected void setImageResource(@NotNull Component component, @NotNull Resource toBeWrapped, @NotNull List<String> hiddenProperties) {
         this.toBeWrapped = toBeWrapped;
         this.component = component;
-        this.hiddenProperties = new ArrayList<String>(hiddenProperties);
+        this.hiddenProperties = new ArrayList<>(hiddenProperties);
+        this.inheritedProperties = new HashMap<>();
+    }
+
+    protected void setImageResource(@NotNull Component component, @NotNull Resource toBeWrapped, @NotNull List<String> hiddenProperties, @NotNull Map<String, String> inheritedProperties) {
```
Comment (vladbailescu) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/AbstractImageDelegatingModel.java:56]: Good opportunity to add missing javadoc to both the previous method and this.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/TeaserResourceWrapper.java (hunk 2 above)
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/TeaserResourceWrapper.java:33]: how about renaming the class `CoreResourceWrapper`? Would it address your comment?

File: content/src/content/jcr_root/apps/core/wcm/components/list/v3/list/item.html [hunk 5]
```diff
@@ -14,7 +14,7 @@
   ~ limitations under the License.
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/-->
 <template data-sly-template.item="${@ list, item}">
-    <article>
+    <article data-sly-test="${!list.displayItemAsTeaser}">
```
[the remark this answers is not in the document]
Comment (jckautzmann) [reply content/src/content/jcr_root/apps/core/wcm/components/list/v3/list/item.html:17]: Are you suggesting to split the template in 2 templates within `item.html` or introducing a new file `teaser.html` containing only one template with the teaser rendering?

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v2/PageListItemImpl.java (hunk 1 above)
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v2/PageListItemImpl.java]: Addressed in https://github.com/adobe/aem-core-wcm-components/pull/1574/commits/6fa9ea079386bd95a8bad46ca676b62ee4d5bec2

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/TeaserResourceWrapper.java (hunk 2 above)
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/TeaserResourceWrapper.java]: Addressed in https://github.com/adobe/aem-core-wcm-components/pull/1574/commits/6fa9ea079386bd95a8bad46ca676b62ee4d5bec2

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/TeaserResourceWrapper.java (hunk 3 above)
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/TeaserResourceWrapper.java]: Addressed in https://github.com/adobe/aem-core-wcm-components/pull/1574/commits/6fa9ea079386bd95a8bad46ca676b62ee4d5bec2

File: content/src/content/jcr_root/apps/core/wcm/components/list/v3/list/item.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (jckautzmann) [reply content/src/content/jcr_root/apps/core/wcm/components/list/v3/list/item.html]: I went ahead and introduced a new file `teaser.html` containing only one template with the teaser rendering.
See https://github.com/adobe/aem-core-wcm-components/pull/1574/commits/6fa9ea079386bd95a8bad46ca676b62ee4d5bec2

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/AbstractImageDelegatingModel.java (hunk 4 above)
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/AbstractImageDelegatingModel.java]: Addressed in 6fa9ea0

**Comment by jckautzmann** [comment]
@vladbailescu I improved the javadocs in https://github.com/adobe/aem-core-wcm-components/pull/1574/commits/f6a219f8f837352bfe20c878bb629d31de1fbe53

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/CoreResourceWrapper.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+import org.apache.sling.api.resource.ResourceWrapper;
+import org.apache.sling.api.resource.ValueMap;
+import org.apache.sling.api.wrappers.ValueMapDecorator;
+import org.apache.sling.models.annotations.Exporter;
+import org.jetbrains.annotations.NotNull;
+
+import com.adobe.cq.export.json.ExporterConstants;
+
+@Exporter(name = ExporterConstants.SLING_MODEL_EXPORTER_NAME , extensions = ExporterConstants.SLING_MODEL_EXTENSION)
+public class CoreResourceWrapper extends ResourceWrapper {
```
Comment (bpauli) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/CoreResourceWrapper.java:33]: This wrapper introduces a lot of duplication from the ImageResourceWrapper. Can it be merged into one?

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/CoreResourceWrapper.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+    public CoreResourceWrapper(@NotNull Resource resource, @NotNull String overriddenResourceType, @NotNull Map<String, String> overriddenProperties) {
+        super(resource);
+        if (StringUtils.isEmpty(overriddenResourceType)) {
+            throw new IllegalArgumentException("The " + CoreResourceWrapper.class.getName() + " needs to override the resource type of " +
+                    "the wrapped resource, but the resourceType argument was null or empty.");
+        }
+        this.overriddenResourceType = overriddenResourceType;
+        valueMap = new ValueMapDecorator(new HashMap<>(resource.getValueMap()));
+        valueMap.put(ResourceResolver.PROPERTY_RESOURCE_TYPE, overriddenResourceType);
```
Comment (bpauli) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/CoreResourceWrapper.java:46]: This logic can be replaced with a org.apache.sling.api.wrappers.CompositeValueMap.

In general I would do this already in the component so the overridden and hidden properties collections get obsolete.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/util/ComponentUtils.java [hunk 8]
```diff
@@ -192,4 +194,14 @@ private static String generateId(@NotNull final Resource resource,
...
-}
\ No newline at end of file
+
+    /**
+     * Returns the resource holding the properties of the featured image of the page.
+     *
+     * @param page the page
+     * @return the featured image resource
+     */
+    public static Resource getFeaturedImage(@NotNull Page page) {
```
Comment (bpauli) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/util/ComponentUtils.java:204]: add @Nullable

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/AbstractImageDelegatingModel.java [hunk 9]
```diff
@@ -42,13 +44,39 @@
...
+     * @param component The component that holds the property of the image to delegate to.
+     * @param toBeWrapped The resource to be wrapped into an image resource.
+     * @param hiddenProperties The properties that are removed from the wrapped image resource.
+     */
     protected void setImageResource(@NotNull Component component, @NotNull Resource toBeWrapped, @NotNull List<String> hiddenProperties) {
         this.toBeWrapped = toBeWrapped;
         this.component = component;
-        this.hiddenProperties = new ArrayList<String>(hiddenProperties);
+        this.hiddenProperties = new ArrayList<>(hiddenProperties);
+        this.overriddenProperties = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/AbstractImageDelegatingModel.java]: I refactored the code in https://github.com/adobe/aem-core-wcm-components/pull/1574/commits/50dfbb9065dd5d10eb4d6cf268d757815b4a6543

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/AbstractImageDelegatingModel.java [hunk 10]
```diff
@@ -42,13 +44,39 @@
...
+
+    /**
+     * Sets the resource that is used for the image rendering of the delegating component.
+     *
+     * @param component The component that holds the property of the image to delegate to.
+     * @param toBeWrapped The resource to be wrapped into an image resource.
+     * @param hiddenProperties The properties that are removed from the wrapped image resource.
+     * @param overriddenProperties The properties that are overridden in the wrapped image resource.
+     */
+    protected void setImageResource(@NotNull Component component, @NotNull Resource toBeWrapped, @NotNull List<String> hiddenProperties, @NotNull Map<String, String> overriddenProperties) {
```
[the remark this answers is not in the document]
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/AbstractImageDelegatingModel.java]: I refactored the code in https://github.com/adobe/aem-core-wcm-components/pull/1574/commits/50dfbb9065dd5d10eb4d6cf268d757815b4a6543

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/PageListItemImpl.java [hunk 11]
```diff
@@ -45,7 +45,7 @@
     /**
      * Standard logger.
      */
-    private static final Logger LOGGER = LoggerFactory.getLogger(PageListItemImpl.class);
+    protected static final Logger LOGGER = LoggerFactory.getLogger(PageListItemImpl.class);
```
[the remark this answers is not in the document]
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/PageListItemImpl.java]: I refactored the code in https://github.com/adobe/aem-core-wcm-components/pull/1574/commits/50dfbb9065dd5d10eb4d6cf268d757815b4a6543

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/CoreResourceWrapper.java (hunk 6 above)
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/CoreResourceWrapper.java:35]: I refactored the code in https://github.com/adobe/aem-core-wcm-components/pull/1574/commits/50dfbb9065dd5d10eb4d6cf268d757815b4a6543 and got rid of the `ImageResourceWrapper`

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/util/ComponentUtils.java (hunk 8 above)
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/util/ComponentUtils.java:205]: Addressed in https://github.com/adobe/aem-core-wcm-components/pull/1574/commits/50dfbb9065dd5d10eb4d6cf268d757815b4a6543

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/CoreResourceWrapper.java (hunk 7 above)
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/resource/CoreResourceWrapper.java:53]: `CompositeValueMap` doesn't do exactly what is needed here as it only merges the properties. There are use cases where we only want to remove properties, some where we only want to override properties and some where a combination of those is needed. I'd like to keep both `hiddenProperties` and `overriddenProperties` to have more flexibility.

File: bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/models/v2/TeaserImplTest.java [hunk 12]
```diff
@@ -96,4 +97,18 @@ protected void testTeaserWithActions() {
...
 
+    @Test
+    protected void testEmptyTeaserImageDelegatingToFeaturedImage() {
+        Teaser teaser = getTeaserUnderTest(TEASER_20);
+        Resource imageResource = teaser.getImageResource();
+        ValueMap imageProperties = imageResource.getValueMap();
+        String linkURL = imageProperties.get("linkURL", String.class);
+        String fileReference = imageProperties.get("fileReference", String.class);
+        assertEquals("/content/teasers/jcr:content/cq:featuredimage", imageResource.getPath(), "image resource: path");
+        assertEquals("core/wcm/components/image/v2/image", imageResource.getResourceType(), "image resource: resource type");
```
Comment (vladbailescu) [inline bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/models/v2/TeaserImplTest.java:108]: Shouldn't this be `v3` already?

File: bundles/core/src/test/resources/list/v3/exporter-childrenListType.json [hunk 13]
```diff
@@ -6,61 +6,65 @@
...
+                "valid": true,
+                "url": "/content/list/pages/page_1/page_1_1.html"
             },
+            "description": "Description for Page 1.1",
             "title": "Page 1.1",
+            ":type": "cq:PageContent",
             "dataLayer": {
                 "list-f56abdba21-item-70236eb2fc": {
-                    "@type": "cq:PageContent",
+                    "@type": "cq:Component/item",
```
Comment (vladbailescu) [inline bundles/core/src/test/resources/list/v3/exporter-childrenListType.json:17]: This is weird, isn't it supposed to still be `cq:PageContent`, like the `:type`?

File: bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/models/v2/TeaserImplTest.java (hunk 12 above)
Comment (jckautzmann) [reply bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/models/v2/TeaserImplTest.java]: Addressed in https://github.com/adobe/aem-core-wcm-components/pull/1574/commits/fe69c8521f680287684b16203fca25fb4d822e53

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v2/PageListItemImpl.java [hunk 14]
```diff
@@ -15,21 +15,41 @@
...
+import com.day.cq.commons.jcr.JcrConstants;
 import com.day.cq.wcm.api.Page;
 import com.day.cq.wcm.api.components.Component;
 import com.fasterxml.jackson.annotation.JsonIgnore;
 
+import static com.adobe.cq.wcm.core.components.models.List.PN_TEASER_DELEGATE;
+
 public class PageListItemImpl extends com.adobe.cq.wcm.core.components.internal.models.v1.PageListItemImpl {
 
+    private Component component;
```
[the remark this answers is not in the document]
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v2/PageListItemImpl.java]: Addressed in 50dfbb9

File: bundles/core/src/test/resources/list/v3/exporter-childrenListType.json (hunk 13 above)
Comment (jckautzmann) [reply bundles/core/src/test/resources/list/v3/exporter-childrenListType.json:17]: This doesn't seem to be related to the changes in this PR. I created https://github.com/adobe/aem-core-wcm-components/issues/1589 to tackle it.

**Review by vladbailescu — APPROVED** [review]

**Review by bpauli — APPROVED** [review]

**Pull request merged**